### PR TITLE
TFS #154577: Implement Todo component

### PR DESCRIPTION
**Pull request by PremnishaTW**
Implement todo component

### CODE REVIEW

File: src/todo/columns.tsx [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import buildColumn from "../refapp-grid/column-builder";
+
+export default function getColumns() {
+  const getDoneActionColumn = () => {
+    return [];
+  };
+
+  const getTodoActionColumn = () => {
+    const fetchEncounterURL = (patientId, encounterId) => {
+      let baseUrl: string = `${location.origin}/openmrs/htmlformentryui/htmlform/viewEncounterWithHtmlForm.page?`;
```
Comment (rrameshbtech) [inline src/todo/columns.tsx]:
```suggestion
      const baseUrl: string = `/openmrs/htmlformentryui/htmlform/viewEncounterWithHtmlForm.page?`;
```

- use constant where ever it is possible.
- use relative path

File: src/todo/columns.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    };
+    const getTodoActionButton = todo => {
+      switch (todo.type) {
+        case "PRINT_CONSENT":
+          return (
+            <a
+              title="Basic Service Plan Encounter URL"
+              href={fetchEncounterURL(todo.patient.id, todo.encounterId)}
+            >
+              <button className="task button small-button">Print</button>
```
Comment (rrameshbtech) [inline src/todo/columns.tsx]: Please use <Trans> for text. So that it will support translation.

File: src/todo/columns.tsx [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      let baseUrl: string = `${location.origin}/openmrs/htmlformentryui/htmlform/viewEncounterWithHtmlForm.page?`;
+
+      return `${baseUrl}patientId=${patientId}&encounter=${encounterId}`;
+    };
+    const getTodoActionButton = todo => {
+      switch (todo.type) {
+        case "PRINT_CONSENT":
+          return (
+            <a
+              title="Basic Service Plan Encounter URL"
```
Comment (rrameshbtech) [inline src/todo/columns.tsx]: Can we have translated content here?

File: src/todo/todo.component.test.tsx [hunk 4]
```diff
@@ -1,22 +1,212 @@
...
   });
 
-  it(`renders Todo without dying`, () => {
-    render(<Todo showMessage={jest.fn()} {...commonWidgetProps} />);
+  afterAll(() => {
+    console.error = originalError;
   });
 
-  it(`should render header component`, () => {
+  it(`should show error message for sourceApi errors`, done => {
```
Comment (rrameshbtech) [inline src/todo/todo.component.test.tsx]:
```suggestion
  it(`should show error message for source api fails`, done => {
```

File: src/todo/todo.component.test.tsx [hunk 5]
```diff
@@ -1,22 +1,212 @@
...
+    const { getByText } = render(
+      <Todo
+        sourceApi=""
+        showMessage={jest.fn}
+        title={componentTitle}
+        {...commonWidgetProps}
+      />
+    );
+
+    expect(getByText("Loading...")).toBeTruthy;
```
Comment (rrameshbtech) [inline src/todo/todo.component.test.tsx]:
```suggestion
    expect(getByText("Loading...")).toBeTruthy();
```

File: src/todo/todo.component.test.tsx [hunk 6]
```diff
@@ -1,22 +1,212 @@
...
+        sourceApi={mockUrl}
+      />
+    );
+
+    waitForElement(() => getByText(componentTitle)).then(() => {
+      done();
+    });
+  });
+
+  it(`should have url for view all`, done => {
```
Comment (rrameshbtech) [inline src/todo/todo.component.test.tsx]:
```suggestion
  it(`should contain view all which redirects to url given in config`, done => {
```

File: src/todo/todo.component.test.tsx [hunk 7]
```diff
@@ -1,22 +1,212 @@
...
+        "task button small-button"
+      );
+      expect(document.querySelector("a").getAttribute("href")).toContain(
+        "openmrs/htmlformentryui/htmlform/viewEncounterWithHtmlForm.page?patientId=14&encounter=37"
+      );
+      done();
+    });
+  });
+
+  it(`should sort Todo's by date`, done => {
```
Comment (rrameshbtech) [inline src/todo/todo.component.test.tsx]:
```suggestion
  it(`should sort Todos by todo created date`, done => {
```

File: src/todo/todo.component.tsx [hunk 8]
```diff
@@ -1,23 +1,113 @@
...
+import resources from "./translations/index";
 
 export default function Todo(props: TodoProps) {
   initI18n(resources, props.locale, useEffect);
+  const [todos, setTodos] = useState(null);
+  const [currentRefreshInterval, setCurrentRefreshInterval] = useState(null);
+  const [loadingStatus, setLoadingStatus] = useState(LoadingStatus.Loading);
+
+  const secondInMilliSeconds = 1000;
+  const source = "/frontend/mockTodo.json";
```
Comment (rrameshbtech) [inline src/todo/todo.component.tsx]: Please remove mock data consumption.

File: src/constants.json [hunk 9]
```diff
@@ -9,5 +9,10 @@
   },
   "charts": {
     "notFoundMessage": "Chart not found"
+  },
+  "todo": {
+    "sortBy": "dateCreated",
```
[the remark this answers is not in the document]
Comment (PremnishaTW) [reply src/constants.json]: Done https://github.com/openmrs/openmrs-esm-user-dashboard-widgets/pull/15/commits/e389ac964206c6b482c1d7a4eb37eddccb2f89e9

File: src/todo/columns.tsx (hunk 2 above)
Comment (PremnishaTW) [reply src/todo/columns.tsx]: Done https://github.com/openmrs/openmrs-esm-user-dashboard-widgets/pull/15/commits/35eba98c8075fafe5402a2c0ecbe61498d053f5b#diff-178c5e4eae7d30925dec99b3f1d5e8b8R25

File: src/todo/columns.tsx (hunk 3 above)
Comment (PremnishaTW) [reply src/todo/columns.tsx]: Removed title will add a backlog card for this

File: src/todo/todo.component.tsx (hunk 8 above)
Comment (PremnishaTW) [reply src/todo/todo.component.tsx]: Done https://github.com/openmrs/openmrs-esm-user-dashboard-widgets/pull/15/commits/c73c39c43edf84f3a53e58f34ae938f6bbb19d89

File: src/todo/todo.component.tsx [hunk 10]
```diff
@@ -1,23 +1,113 @@
...
+  };
+
+  const sortTodos = fetchTodos => {
+    const compareTodo = (current, next) =>
+      current[constants.SORT_BY] - next[constants.SORT_BY];
+
+    return fetchTodos.sort(compareTodo);
+  };
+
+  const getTodoByLimit = () => {
```
Comment (rrameshbtech) [inline src/todo/todo.component.tsx]:
```suggestion
  const limitListByCount = (items, limit) => {
```

File: src/todo/todo.component.tsx [hunk 11]
```diff
@@ -1,23 +1,113 @@
...
+
+  const showGrid = () => {
+    return (
+      <div>
+        <WidgetHeader
+          title={props.title}
+          icon="svg-icon icon-todo"
+        ></WidgetHeader>
+        <RefAppGrid
+          data={getTodoByLimit()}
```
Comment (rrameshbtech) [inline src/todo/todo.component.tsx]:
```suggestion
          data={limitListByCount(todos, limit)}
```

Comment (PremnishaTW) [reply src/todo/todo.component.tsx]: done https://github.com/openmrs/openmrs-esm-user-dashboard-widgets/pull/15/commits/e878639e2783f7f7c9535109d61458b628050d92

File: src/todo/todo.component.tsx (hunk 10 above)
Comment (PremnishaTW) [reply src/todo/todo.component.tsx]: done https://github.com/openmrs/openmrs-esm-user-dashboard-widgets/pull/15/commits/e878639e2783f7f7c9535109d61458b628050d92

**Pull request merged**